### ISSUE TITLE
Severe performance regression with software decoding

### ISSUE THREAD

**Issue by versedwildcat** [mpv-android/mpv-android#508]
### Important information
**Manufacturer:** Samsung 
**Model:** [Galaxy S8+ G955F](https://www.gsmarena.com/samsung_galaxy_s8+-8523.php#g955f)
**OS:** Stock Android 9
**Chipset:** Exynos 8895 Octa
**mpv-android version:** 2022-03-06
**Which version of mpv-android introduced the problem (if known):**
Software Playback broke somewhere between 2021-10-28 and [2022-01-02](https://kitsunemimi.pw/tmp/mpv-android-2022-01-02.apk)

### Description
Playback with Software Decoding is completely broken with most common codecs on my Exynos Galaxy S8+
I have run some tests using video samples _(mostly)_ from https://test-videos.co.uk/ and https://kodi.wiki/view/Samples
First, I used MPV-android out of the box, meaning Hardware decoding was enabled.
Then, I toggled off HWDEC and tried again
Here are the results:
| Test 1 | HWDEC | SWDEC                                                                               |
|-------------|-------|-------------------------------------------------------------------------------------|
| x264        | works | Video plays at normal speed <br>(1sec == 1sec IRL)<br>Extremely low FPS (Slideshow) |
| VP9         | works | Plays at normal speed <br>Extremely low FPS (Slideshow)                             |
| HEVC        | works | Slideshow, plays at slow speed <br>(1sec == 4sec IRL)                               |
| HEVC High10 | works | Black Screen, plays at slow speed                                                   |

|  Test 2    | HWDEC               | SWDEC                                                                              |
|------------------------------|---------------------|------------------------------------------------------------------------------------|
| HEVC High10 w/ ASS Subtitles | works               | Black Screen, plays at slow speed, <br>Broken sound, subtitles don't appear at all |
| x264 High10                  | Falls Back to SWDEC | Black Screen, extremely slow playback, broken sound                                |
| x264 High444p                | Falls Back to SWDEC | Black Screen, extremely slow playback, broken sound                                |
| AV1                          | Falls Back to SWDEC | works                                                                              |

It's pretty funny how AV1 is still the odd one out, by being the only codec the works.

### Log output
Here's the [Tests logcat.zip](https://github.com/mpv-android/mpv-android/files/8233069/Tests.logcat.zip) of the various tests, from `adb logcat -s 'mpv:*' '*:F'`
### Additional information
As a side effect, switching between hardware and software decoding during playback is completely broken.
All files play fine on MPV on desktop 
All files play fine (except ASS subtitles) on VLC-android, both with Hardware Acceleration set to "Disabled" and "Full Acceleration"
Probably related to: #484 
![black_screen](https://user-images.githubusercontent.com/96885152/157893707-5346c904-ced2-4f30-8d49-700e7c8aef4f.jpeg)

**Comment by sfan5** [comment]
Using a standalone benchmark I observe that raw decoding performance regressed a tiny bit: (but that won't be it)
```
-- 2022-03-06
bench: utime=122.588s stime=3.472s rtime=23.789s
bench: maxrss=132616kB
bench: utime=124.076s stime=3.172s rtime=21.703s
bench: maxrss=132728kB
bench: utime=123.792s stime=3.088s rtime=20.755s
bench: maxrss=132800kB

-- 2021-10-28
bench: utime=117.472s stime=3.164s rtime=20.981s
bench: maxrss=133636kB
bench: utime=118.888s stime=3.100s rtime=19.802s
bench: maxrss=133652kB
bench: utime=118.880s stime=3.256s rtime=19.948s
bench: maxrss=133820kB
```

Instead observing the end result I get these outcomes:
<pre>
new = 2022-03-06; old = 2021-10-28
new ffmpeg + new mpv: bad
old ffmpeg + new mpv: bad
new ffmpeg + old mpv: good
old ffmpeg + old mpv: good
</pre>

That puts the regression somewhere into mpv itself.

**Comment by sfan5** [comment]
Suspect from looking at the log is DR. I had no idea this started working (it's enabled automatically) on Android.
This also explains why it doesn't affect dav1d, its external to ffmpeg so can't use DR.

Bisected libmpv:
<pre>
2fc327f2fdd48f0cd58a4d4382a7aa57f3fabd77 is the first bad commit
commit 2fc327f2fdd48f0cd58a4d4382a7aa57f3fabd77
Author: Philip Langdale <philipl@overt.org>
Date:   Sat Dec 4 11:42:16 2021 -0800

    vo_gpu: opengl: fix OpenGL ES version and extension handling
    
    Some of the extension declarations did not include the ES version where
    they became core functionality, and in some of these cases, there was
    never actually an ES extension - it first appeared in core. We also had
    a number of buggy version checks where ES versions were compared
    against required desktop GL versions.

 video/out/opengl/common.c | 15 ++++++++++++++-
 video/out/opengl/ra_gl.c  | 15 +++++++++++----
 2 files changed, 25 insertions(+), 5 deletions(-)
</pre>

<b>As a workaround the following in your config should fix performance: </b> `vd-lavc-dr=no`

**Comment by sfan5** [comment]
By the way [my phone](https://www.gsmarena.com/samsung_galaxy_a40-9642.php) has an Exynos too, not sure if it affects other SoCs but going by #507 it might.

**Comment by versedwildcat** [comment]
Great work!
I added `vd-lavc-dr=no` to mpv.conf and performance did definitely improve
The tests are of course using software decoding. I'll upload a log file in asap
|Codec|Result|
|------|------|
|x264|Works|
|VP9|Works|
|HEVC|Works|
|HEVC High10|Normal Video Speed, Black Screen|
|HEVC High10 & ASS Subs|Normal Video Speed, Black Screen, Sound normal, subtitles don't appear at all|
|x264 High10|Normal Video Speed, Black Screen, Sound normal|
|x264 High444p|Normal Video Speed, Black Screen, Sound normal|

**Comment by versedwildcat** [comment]
Here's the logcat of the [working ones.txt](https://github.com/mpv-android/mpv-android/files/8234919/working.ones.txt) and here are the logs of the videos that only show a  [black screen.txt](https://github.com/mpv-android/mpv-android/files/8234918/black.screens.txt).
Could this be an unrelated issue?

**Comment by sfan5** [comment]
10-bit not rendering anything is "normal" (read: a know issue) at least with Mali GPUs, see #198.
The current workaround is to put `vf=format=yuv420p` into your config.

**Comment by sfan5** [comment]
fixed by ffd1e8f5e0bb40a9014481a41ae9bad48866381e

**Comment by vegedb** [comment]
> fixed by [ffd1e8f](https://github.com/mpv-android/mpv-android/commit/ffd1e8f5e0bb40a9014481a41ae9bad48866381e)

Will there be a mini update for this?

**Comment by sfan5** [comment]
Yes, I have a few other things I want to do first.
